feat: Add port configuration support for build command

dougwithseismic opened the pull request:
## Description
Following up on https://github.com/mastra-ai/mastra/pull/3025 ...

This PR adds port configuration support to the `mastra build` command, making it consistent with the existing port configuration in `mastra dev`. It provides a unified way to configure server ports across development and production environments.

## Changes
- Added `--port` flag to `mastra build` command
- Enhanced `createNodeServer` to accept port configuration
- Updated build process to properly handle port configuration
- Improved port resolution logic in production
- Updated port configuration documentation

## Implementation Details
The port is now determined in this order:
1. Command line argument (--port)
2. Environment variable (process.env.PORT)
3. Default value (4111)

## Testing
- [x] Tested `mastra build --port 3000`
- [x] Verified port configuration through environment variables
- [x] Confirmed default port fallback works
- [x] Tested in development with `mastra dev --port 3000`
- [x] Verified production builds maintain port configuration

## Documentation
Updated documentation to reflect new port configuration options in build commands.

## Breaking Changes
None. This is a backwards-compatible enhancement.

dougwithseismic committed:
feat: Add port configuration support for build command - Add port CLI option, update BuildBundler, and documentation

dougwithseismic committed:
refactor: remove redundant port configuration in dev.entry.js

dougwithseismic committed:
feat: add port configuration to bundler with minimal changes

dougwithseismic committed:
refactor: remove unused rollup import

dougwithseismic committed:
fix: use replaceAll to handle multiple instances of process.env.PORT

dougwithseismic committed:
refactor: separate bundleLocation and options parameters in _bundle method

dougwithseismic committed:
refactor: simplify port configuration by mutating serverFile directly

dougwithseismic commented:
Not 100% confident I picked up everywhere here - am I missing anything else?

dougwithseismic committed:
Merge branch 'main' into port-configuration

wardpeet committed:
fixes

wardpeet commented:
I've updated your PR a little, we should only look at port in dev, build should not. For build you should set the env PORT yourself before running it

wardpeet committed:
fix docs

dougwithseismic commented:
So we don't want support custom ports for when Mastra is part of more
intricate projects using the same port env?

E.g.  https://github.com/dougwithseismic/withseismic-mcp/tree/main

On Fri, 21 Mar 2025, 11:42 Ward Peeters, ***@***.***> wrote:

> I've updated your PR a little, we should only look at port in dev, build
> should not. For build you should set the env PORT yourself before running it
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2742985808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALSIKRBI7WGWLVUGDFANKKT2VPUKFAVCNFSM6AAAAABZJRCI3SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBSHE4DKOBQHA>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
> [image: wardpeet]*wardpeet* left a comment (mastra-ai/mastra#3044)
> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2742985808>
>
> I've updated your PR a little, we should only look at port in dev, build
> should not. For build you should set the env PORT yourself before running it
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2742985808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALSIKRBI7WGWLVUGDFANKKT2VPUKFAVCNFSM6AAAAABZJRCI3SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBSHE4DKOBQHA>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

dougwithseismic commented:
Maybe exposing hono is a good idea instead.

On Fri, 21 Mar 2025, 11:45 Doug Silkstone, ***@***.***> wrote:

> So we don't want support custom ports for when Mastra is part of more
> intricate projects using the same port env?
>
> E.g.  https://github.com/dougwithseismic/withseismic-mcp/tree/main
>
> On Fri, 21 Mar 2025, 11:42 Ward Peeters, ***@***.***> wrote:
>
>> I've updated your PR a little, we should only look at port in dev, build
>> should not. For build you should set the env PORT yourself before running it
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2742985808>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ALSIKRBI7WGWLVUGDFANKKT2VPUKFAVCNFSM6AAAAABZJRCI3SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBSHE4DKOBQHA>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>> [image: wardpeet]*wardpeet* left a comment (mastra-ai/mastra#3044)
>> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2742985808>
>>
>> I've updated your PR a little, we should only look at port in dev, build
>> should not. For build you should set the env PORT yourself before running it
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2742985808>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ALSIKRBI7WGWLVUGDFANKKT2VPUKFAVCNFSM6AAAAABZJRCI3SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBSHE4DKOBQHA>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
>

wardpeet commented:
we do but why can't you set the PORT?

wardpeet committed:
remove from bundler

dougwithseismic commented:
My thoughts are itd be nice because port might be used for another service.
I don't think it's great dev experience for Mastra to hijack the main PORT
env without offering a mechanism to override it.

If you're allowing port to be overwritten in dev then why are we
constrained to a hardcoded env var for builds?

Just my thoughts as I integrate mastra into a server that will handle auth
etc.


On Fri, 21 Mar 2025, 12:22 Ward Peeters, ***@***.***> wrote:

> we do but why can't you set the PORT?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2743074916>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALSIKRFLNXYGMTH3SOSBZWD2VPY7VAVCNFSM6AAAAABZJRCI3SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBTGA3TIOJRGY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
> [image: wardpeet]*wardpeet* left a comment (mastra-ai/mastra#3044)
> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2743074916>
>
> we do but why can't you set the PORT?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mastra-ai/mastra/pull/3044#issuecomment-2743074916>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALSIKRFLNXYGMTH3SOSBZWD2VPY7VAVCNFSM6AAAAABZJRCI3SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBTGA3TIOJRGY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

wardpeet commented:
Ok  I see, so your basically running 2 web servers and talking them using fetch calls In the same docker container or application. That's an odd use case, I guess what you're actually looking for is how can I use the same handlers or extend our hono server with yours? Am I correct?

wardpeet commented:
@dougwithseismic I'm going to close this PR as we now have the server key in configuration that we can use to define a port on dev and build. This can be set to a custom env var.

```
new Mastra({
  server: {
    port: process.env.MY_CUSTOM_PORT_KEY
  }
})
```
